### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by billh0420**
New game Gin Rummy

### CODE REVIEW

**Comment by daochenzha** [comment]
@billh0420 Hi, we have received the request. We need more time to work on the code review since there are lots of files. Thank you for contributing this new game!

**Comment by daochenzha** [comment]
@billh0420 Hi, it would be nice if you could also provide some testing codes as in https://github.com/datamllab/rlcard/tree/dev/tests Testing will be of great help in future maintenance.

**Comment by daochenzha** [comment]
@billh0420 Hi, thanks again for the contribution! I have merged the request but made some slight modifications.
1. I replace the f-string with traditional format to support python 3.5
2. I move some files into https://github.com/datamllab/rlcard/tree/dev/rlcard/games/gin_rummy/utils to make the structure clearer.
2. Remove 'print_state' etc. since we use human agent to do the text interface.
3. remove `z_test.py`, which seems to be testing file.
4. Modify the example a little bit to adapt the interfaces in Dev branch

Overall, it is a nice implementation. However, there are still some issues that I hope you can take a look:
1. The Game engine seems not to be well wrapped. Specifically, I feel like the Scorer should be wrapped inside judge. Currently, i see that the Scorer, Judger, Player are imported in environment. This may make the code structure not very clear. The original goal of env is just to wrap the game engine with state and action representations. Thus, I think should be wrapped in the game.
2. I found there is a HighLowAgent in the gin rummy. It would be better to put this rule agent under `/models`. An example is https://github.com/datamllab/rlcard/blob/dev/rlcard/models/leducholdem_rule_models.py
3. The testing file only provides very limited coverage, it would be nice to enrich the testing scripts and also write a test for environment like https://github.com/datamllab/rlcard/blob/master/tests/envs/test_leducholdem_env.py
4. More comments are needed. I can roughly follow the codes but still lost in some details, such as in round.py

I wish to make gin_rummy to have same style as the other games in the package, so that we can make things easier for future maintenance. Thanks!

**Comment by billh0420** [comment]
I am looking at item 2 of your request of changes. You asked that the HighLowAgent be put under /models. However, I don't have a model class. I don't envision using the HighLowAgent to compute against the HighLowAgent. The HighLowAgent plays a perfect strategy. I envision using it in place of the RandomAgent in the various training environments.

Is it ok if I just put it in /models under the name of gin_rummy_high_low_agent.py ? It is not a model.

**Comment by daochenzha** [comment]
Hi, there is folder named models, which serves as a model zoo to put some rule model or pertained models. An example is a rule agent for Leduc  https://github.com/datamllab/rlcard/blob/master/rlcard/models/leducholdem_rule_models.py
Then the model should be registered in
https://github.com/datamllab/rlcard/blob/master/rlcard/models/__init__.py
Just like the environment. Then when using models, we can conveniently use model.load

**Comment by daochenzha** [comment]
In the example of Leduc Rule, the only difference between model and agent is that model contains agents in all the seats. This is for the game when different positions need different strategies. In Leduc example, we simply use the same agent for all the seats. I still recommend to wrap your rule agent with a model class, which may be more extendable in the future.

**Comment by billh0420** [comment]
I am looking at the first item 4 where you request that I remove z_test.py which I use for testing.

This is one of my ways to test my code, especially when I know that there is a problem. I would like to be able to keep this somewhere for my use. Do you have a suggested place to keep it or do you want me to just delete it.

For example, test02 shows me how frequent the random agents will end the game without either going out by knocking or ginning. The result is about 5 times out of a thousand. Thus, one of my goal for training would be for the trained agent to go out about 900 times out of 1000 since it is rare that neither player knocks or gins. This is not really something for a unit test: it is something to give me a feeling for what is going on.

**Comment by daochenzha** [comment]
@billh0420 I understand that we need a script for testing to know what is going on. However, I do not think it is appropriate to put this file in the game engine. It may be clear to you what the testing script is about but could be confusing to others. I would suggest keeping this testing script privately in your place and use it when debugging. Another way to could be to transform this z_test.py into some sort of integration test and put it in `/tests`

**Comment by billh0420** [comment]
This is related to each developer/user having their own "private" files somewhere.

I am fighting on how to allow different settings to be applied to the running of code in rlcard. Right now I am just modifying the python file to reflect the settings that I want.  For example, in rlcard.examples.ginrummy_dqn.py, I modified line 62 to read:

   episode_num = 1000  # mahjong_dqn has 100000

This is the simplest example of a user modification.

My High-Low settings for Gin-Rummy is a major reconfiguration of the basic Gin Rummy game. I put the code to handle this reconfiguration in that same file. However, it seems that this isn't what you want.

First, I could just drop the High_Low idea since it is just something I want to use to learn how to do reinforcement learning and has nothing to do with Gin Rummy. I could just move the High_Low code to my private project folder to do my experimentation there and leave rlcard intact.

Second, there still remains other settings that are helpful to run Gin Rummy code for reinforcement learning. In particular, the different methods of getting the payoffs. I guess I could settle on one payoff method and just use it. Again, I could document how the user could change the payoff calculation in their own "private" project using rlcard.

In summary, I will be going to this simplified version of using just one method that I think is appropriate. I will delete all references to High-Low except putting it into the documentation on Gin Rummy on how to reconfigure Gin Rummy to run in High-Low mode.

My Question: Does this make sense and is it ok with you? Or, do you have other suggestions?

**Comment by daochenzha** [comment]
@billh0420 Thanks for the comment! It is very nice to make the game configurable. I notice that currently the choices such as the payoffs are written in the examples. I would suggest moving the two payoffs functions into  https://github.com/datamllab/rlcard/blob/master/rlcard/envs/ginrummy.py And use the `config` (a dictionary) passed to the environment for this configuration. For example, we can define a `payoff` field in `config` and explicitly document it. I think there are mainly two benefits to doing so. First, we can make the example shorter and easier for others. Second, the other developers can use your configuration easily by just passing a value to the environment (otherwise they need to copy-paste your choices every time when they want to use your payoff calculation).

Both our ideas of the High_Low agent will work. In your idea, you treat it as a reconfiguration of the environment. In our idea, we treat the High_Low agent as a model (that is separated from the environment. Since all other environments put rule-agent into `/models`, we would suggest you follow this style since I believe no one would like to use a library that is not inherently consistent. For your idea of making High_Low agent as a reconfiguration of the environments. We actually have a design similar to your idea! We have a `single_agent_mode` in Leduc Holdem, see https://github.com/datamllab/rlcard/blob/master/rlcard/envs/env.py#L45 In this mode, the other players will use pre-trained models (or rule agents), and the environment essentially becomes a single-agent environment. I think this should be similar to your idea of High-Low mode? If it is the case, do you think it would be better to wrap your high_low idea into this interface?

Let me know if I miss something or misunderstand your questions.

**Comment by billh0420** [comment]
1) My basic problem is that the user of rlcard is going to have to make modifications somewhere. But it seems that there is no where for the user to store those modifications. In fact, it would be impossible to do that if there is a requirement that rlcard doesn't change any files.

My solution to this problem is the following. The user creates his own project based on GitHub rlcard. He should then add folders outside of rlcard to do his modifications that will invoke rlcard code. He can periodically pull to keep his rlcard stuff up to date without affecting his modifications. I suspect that this is being done now. The difference is that I, as a developer, would do the same thing: create a user rlcard project but also keep my dev rlcard project. I would do my experimentation in my user rlcard project. I would to my development work in my dev rlcard project.

Question 1: You really didn't address this issue of where the user or dev stores his "configuration" that he is currently using. It seems that you are saying that the developer provides useful canned configurations for the user to employ. But, I think there are too many variations to cover all configurations that the user might want to use. I gave the example of "episode_num = 1000" where who knows how many episodes the user wants to run. This has to be stored somewhere, preferably not somewhere underneath the top level folder rlcard.

2) My settings or configuration file for Gin Rummy would have the following parameters:

          scorer_name
          dealer_for_round
          stockpile_dead_card_count
          going_out_deadwood_count
          max_drawn_card_count
          is_allowed_knock
          is_allowed_gin
          is_allowed_pick_up_discard
          is_allowed_to_discard_picked_up_card
          is_always_knock
          is_south_never_knocks

The number of possible variations is large. I could decide what I think is useful, but the point of the configuration file is to let the user decide. If there are only canned configuration options, then the user choices are unnecessarily limited since he could make his own configuration and use it.

**Comment by daochenzha** [comment]
I agree with you. It seems better to have a file in each environment to decide some parameters for each game so that developers can easily configure the settings. What I can imagine is a file named `config.py` or `setting.py` under the folder of each game (just like your setting.py). We can keep the `choice` in the example.

For the `choice_get_payoff `, I still hold my opinions that it would be better to wrap it into the environment rather than expose it in the example file. I mean we can keep using `choice_get_payoff`, but put the function `def get_payoff_choice1` into https://github.com/datamllab/rlcard/blob/master/rlcard/envs/ginrummy.py 

For the HighLowAgent, we still use it. We just move https://github.com/datamllab/rlcard/blob/master/rlcard/games/gin_rummy/agents.py to https://github.com/datamllab/rlcard/tree/master/rlcard/models and wrap it with a model class just like other games.

And for `choice` and `choice_get_payoff `, we can use string instead of the integer values, which is more meaningful. We also add more comments into the example to explain what are choices.

Do these make sense to you or do you have other suggestions?

**Comment by billh0420** [comment]
I am trying to push my changes and I get the following error message:

remote: error: File examples/models/gin_rummy_nfsp/model.data-00000-of-00001 is 373.39 MB; this exceeds GitHub's file size limit of 100.00 MB  

I deleted the files are created in examples/models by running the example programs. But that did not help. I googled this problem and it seems that the solution is to undo commits until you get to the time that the file is created. I think for me this is a ways back. I will continue to look for a solution to do a push.

I am pointing this out in case you are not aware of it. It seems that you have to do something on your end to tell git to ignore files in examples/models (and probably in example/experiments).

**Comment by daochenzha** [comment]
Thanks for letting us know. In your case simply deleting the file and commit will not help because the file will be the git log. The simplest way to recover the situation is to clone the original project again and modify the files again. You may also try to use `git revert` to revert to original commit, and use `git push -f` to do a force push.

**Comment by billh0420** [comment]
I pushed my changes to my dev area. There a message there to compare and pull. But it looks like the changes were uploaded to my dev account. Is this message for me or for you? If it is for you, I am not completely done yet, so don't pull to the master of dev if that is what the message is saying.

**Comment by daochenzha** [comment]
It should be on your side. You need to first make your local dev up-to-date before sending the pull request.

**Comment by billh0420** [comment]
I am ready to submit my changes to the Gin Rummy program.

Do I need to ask you or do I just submit the changes ?

**Comment by daochenzha** [comment]
Please send a pull request to dev branch. I will review the code. Thanks.
